src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -122,6 +124,7 @@ protected ShardIterator shards(ClusterState clusterState, InternalRequest reques
 
         long[] preVersions = new long[request.items().length];
         VersionType[] preVersionTypes = new VersionType[request.items().length];
+        Translog.Location location = null;
         for (int requestIndex = 0; requestIndex < request.items().length; requestIndex++) {
```

Comment:
I will give the example again. Current code:
    
    Translog.Location location = null;
    for (loop) {
      location = locationToSync(...);
    }
    processAfter(location);

Much better code:

    Translog.Location maxLocation = null;
    for (loop) {
      location = locationToSync(...);
      // we always get the same log file, and we are append only
      assert maxLocation == null || (maxLocation.file == location.file && maxLocation.pointer > location.pointer);
      maxLocation = location;
    }
    processAfter(maxLocation);

Simply changing variable names and adding comment/assert makes it much more readable. And i do not trust that the transaction log file is not changing. There is too much concurrency and crazy shit here for me to know that. As long as there is stuff like jsr166e package in the codebase, its clear concurrency cannot be trusted.